### ISSUE TITLE
jsp taglib not ready for tag pooling (cfr. jasper2)?

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#290]
**Reported by @frederikhabils on 19 Sep 2003 18:48 UTC**
this started out as a mail with a simple question : what does function
cleanState() do. INstead it turned into the question:

    is makumba ready for tagpooling?

To the best of my understanding

- jasper2 and other jsp compilers, may choose to use tag pools
- tag pools basically mean that tags are reused, instead of being
garbage-collected, after one use.
- so as taglib programmer, you have to ensure that the state of the tag
instance, is 'clean' at the beginning of it's use
- and you do that by cleaning up at the end of each use
- for which the jsp container calls the function Tag.release() after each use of
the tag. 

(see p. 536  = p 10 of 46 in the pdf) 
http://developer.java.sun.com/developer/Books/javaserverpages/projsp/Chap12.pdf 

Now, what worries me pretty much is

- we are using jasper1 in production, but we want to move to jasper2 (bug 184)
- we use jasper2 in parade, i think, and in bundles
- none of makumba Tags have a release() 
- a few makumba *Tag.java have cleanState(), which is doing some cleanup.
- but these cleanState() functions are just sitting there, they are never
called, AFAI can tell

C:\dev\bundle\sources\makumba\classes>grep -r cleanState *
org/makumba/view/jsptaglib/IfTag.java:  public void cleanState()
org/makumba/view/jsptaglib/IfTag.java:    super.cleanState();
org/makumba/view/jsptaglib/MakumbaTag.java:  public void cleanState()
org/makumba/view/jsptaglib/QueryTag.java:  public void cleanState()
org/makumba/view/jsptaglib/QueryTag.java:    super.cleanState();
org/makumba/view/jsptaglib/ValueTag.java:  public void cleanState()
org/makumba/view/jsptaglib/ValueTag.java:    super.cleanState();

[duh] google is pointing back to makumba webpages, for "cleanState" ;-)
http://www.google.be/search?q=cleanState  (makumba listed at #5) 

So, although I haven't seen any errors yet, I'm pretty very worried now...

Migrated-From: http://trac.makumba.org/ticket/583

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 20 Sep 2003 11:51 UTC**
cleanState() should have been called in  MakumbaTag, most probably in release()
i have no idea why that call is missing. it may have been removed by the joining
of some branch... would probably be useful to examine how MakumbaTag evolved on
the branch where this new code was started. no time to do that now

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 20 Sep 2003 18:26 UTC**
i renamed cleanState() to release()
this should be fixed, though it's pretty hard to test, needs lots of accesses,
or a small pool or so

i don't think other tags need a release() but feel free to propose
for example IfTag doesn't need it because expr is a required attribute so it
will be set at reuse, no need to nullify in release()

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 20 Sep 2003 18:30 UTC**
forgot to close
i'll add that yes, cleanState() was especially designed for tag pools

**Comment by ghost** [comment]
**Comment by @frederikhabils on 21 Sep 2003 10:36 UTC**
added to 0.5.9.13

**Comment by ghost** [comment]
**Comment by @frederikhabils on 4 Dec 2003 23:50 UTC**
I reopened this bug, because analysis of bug 654 has shown that my initial
report here has Serious error. 

I made wrong conclusions based on the image here :

(see p. 536  = p 10 of 46 in the pdf) 
http://developer.java.sun.com/developer/Books/javaserverpages/projsp/Chap12.pdf 

It turns out, that it is **NOT** true that release() is called in between uses
of the tag in the tagpool. 

I won't repeat it all here, but I advice CAREFUL reading of 

http://nagoya.apache.org/bugzilla/show_bug.cgi?id=16001
http://www.onjava.com/pub/a/onjava/2001/11/07/jsp12.html
 -> Page 2, the "Tag handler life cycle and instance reuse" section


Bottom line, 
- the doStartTag is pretty much responsible for cleaning the state that might
have remained from previous tag executions.
- our release() methods are mostly useless, and mostly cleaning up attribute
values that don't need cleaning up.

I'll probably work more on this in next days.

**Comment by ghost** [comment]
**Comment by @frederikhabils on 6 Dec 2003 02:45 UTC**
fixed. 

- removed most instances of "release()" (where they were pointless). 
- added the method "initialiseState()" which is called in doStartTag() in
MakumbaTag 


-- committed file list --

/usr/local/cvsroot/makumba/classes/org/makumba/view/jsptaglib/AddTag.java,v
new revision: 2.2; previous revision: 2.1

/usr/local/cvsroot/makumba/classes/org/makumba/view/jsptaglib/FormTagBase.java,v  
new revision: 2.5; previous revision: 2.4

/usr/local/cvsroot/makumba/classes/org/makumba/view/jsptaglib/IfTag.java,v
new revision: 1.5; previous revision: 1.4

/usr/local/cvsroot/makumba/classes/org/makumba/view/jsptaglib/InputTag.java,v  
new revision: 2.3; previous revision: 2.2

/usr/local/cvsroot/makumba/classes/org/makumba/view/jsptaglib/MakumbaTag.java,v
new revision: 2.9; previous revision: 2.8

/usr/local/cvsroot/makumba/classes/org/makumba/view/jsptaglib/NewTag.java,v
new revision: 2.2; previous revision: 2.1

/usr/local/cvsroot/makumba/classes/org/makumba/view/jsptaglib/QueryTag.java,v
new revision: 2.6; previous revision: 2.5

/usr/local/cvsroot/makumba/classes/org/makumba/view/jsptaglib/ValueTag.java,v
new revision: 2.4; previous revision: 2.3

**Comment by ghost** [comment]
**Comment by @frederikhabils on 6 Dec 2003 13:22 UTC**
Cristi wrote by mail (in reaction to cvs-commit of ValueTag) :

> hmmmm
> this doesn't look ok...
> 
> > --- ValueTag.java	20 Sep 2003 18:23:05 -0000	2.3
> > -  /** cleanup the state to make this reusable */
> > -  public void release()
> > -  {
> > -    super.release();
> > -    expr=var=printVar=null;
> > -  }
> 
> expr, var, printVar need to be nullified before the setter methods are 
> called, just in case var and printVar are not indicated now, but were 
> indicated last time this ValueTag object was used... to what i know, if 
> they are not indicated, setVar(null) is _not_ called! i remeber vaguely 
> having this problem the first time we experimented with tag reuse.
> 
> i see no code that nulifies them presently?
> 
> maybe this problem is present in more than this file?

I'm quoting from this source (see my reading recommendation earlier)
http://nagoya.apache.org/bugzilla/show_bug.cgi?id=16001

<quote>
If you look through the comments for this "bug", the solution is defined  by the
JSP spec: a tag handler instance can _not_ be reused for occurances of the
corresponding custom action that specify different optional attributes. Hence,
there's no risk that a tag handler that has been used with an optional attribute
is later used without that optional attribute. See this article for details:

 <http://www.onjava.com/pub/a/onjava/2001/11/07/jsp12.html>
  Page 2, the "Tag handler life cycle and instance reuse" section
</quote>

That means (for me) that we don't need to clean up expr, var, printVar. The tag
will only be reused for other instances, where the same attributes are defined
and set before doStartTag().

Also, this is a short summary of best practices around custom tag development,
with respect to life cycle management.

http://jakarta.apache.org/taglibs/guidelines.html

One extension jsp 1.1 -> 1.2 seems to be the new TryCatchFinally interface. If
the tag implements this, you are guaranteed that each usage of tag, is ended
with a call on doFinally() of the tag. 
You could do cleanup or reinitialisation there...

Still, http://jakarta.apache.org/taglibs/guidelines.html seems to suggest to use
doFinally only for resource cleanup, not for re-initialisation, for which they
encourage cleaner coding.

E.g. this recommendation, was one that was/is many times violated by makumba
jsptaglib: "setXXX() methods are meant to be "idempotent"; write them so that
they can be called multiple times with no side effects" 

(mostly simple violations, like : setX (String s) { sb.append(s); }  --> not
"idempotent")

Perhaps, reading further into these guidelines, we may find more things to
refactor in jsptaglib.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 6 Dec 2003 13:55 UTC**
shit, now i realise that the damn image has mislead me too... should we sue? :))

about how tags are reused: i don't care much what jakarta ppl say, as long as
the spec doesn't say it... these days you can't even trust the spec. the spec
seems to mention optional attributes though

idempotence of setters: point taken.

**Comment by ghost** [comment]
**Comment by @frederikhabils on 6 Dec 2003 15:24 UTC**
Cristi wrote by mail (in reaction to cvs-commit of MakumbaTag) :

> in MakumbaTag:
> 
>> -import java.util.Hashtable;
>> +import java.util.*;
> 
> 
> modern coding style requires fully qualified imports.

Ok then... Even if that means listing 5-10 or even 20 used classes?

>> +    // FIXME (fred): what's this? does it do what's expected of it, 
>> knowing that release() does not run after every usage of the tag.
>>
>>      if(findAncestorWithClass(this, MakumbaTag.class)==null)
>>        pageContext.removeAttribute(DB_ATTR);
> 
> this is the db="..." attribute, only permitted for root makumba tags. 
> it's stored in a page attribute for some reason.
> 
> so cleaning it up is needed at every root tag end or, reversely, before 
> every root tag begin, before its setDb() is executed, so i guess 
> setParent is a good place (off the top of my head i think setParent() is 
> executed before setDb(), and i'm not sure about the signature below)
> 
> public void setParent(Tag t){
>  super.setParent(t);
>  if(findAncestorWithClass(this, MakumbaTag.class)==null)
>     pageContext.removeAttribute(DB_ATTR);
> }

I'd be more in favour for not holding the db value in a page attribute, but
rather in the root tag field. The only drawback is that every nested tag that
needs it, will have to go all the way up the "tree" of nested tags, to find it
at the root. (although solutions around that also exist). But given bug 657,
I'll wait altogether with this. 

Apart from that, your setParent suggestion would also do the job, i guess.
Although we must be sure then, that the set page attribute has no unwanted side
effect on other tags at the root, like this : 

<mak:list ...> ... </mak:list> //root 
<mak:attribute ... > // root
<mak:list ...> ... </mak:list> //root 

-> we must make sure that the mak:attribute wouldn't be mislead by the still
existing pageAttribute from the previous mak:list. I don't think there's a
problem with this with today's code, but it's "risky" imo.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 6 Dec 2003 17:14 UTC**
>> modern coding style requires fully qualified imports.
> Ok then... Even if that means listing 5-10 or even 20 used classes?

didn't look into exactly why. but when you've got so many imports it's probably
time to consider refactoring...

> I'd be more in favour for not holding the db value in a page attribute, but
> rather in the root tag field. The only drawback is that every nested tag that
> needs it, will have to go all the way up the "tree" of nested tags, to find it
> at the root. (although solutions around that also exist). 

i don't really remember why an attribute is used, but as far as this fix goes, i
think it's important to have it solved in an easy way in order not to introduce
too many changes or so.

> -> we must make sure that the mak:attribute wouldn't be mislead by the still
> existing pageAttribute from the previous mak:list. I don't think there's a
> problem with this with today's code, but it's "risky" imo.

well, if we'd know that mak:attribute is also affected by this, we'd do the
cleanup there too

**Comment by ghost** [comment]
**Comment by @frederikhabils on 18 Jan 2004 17:03 UTC**
Solution so far had been released as 0.5.9.17. 
Today, a bug was found and fixed -> 0.5.9.18 (should be tested). 

The bug was that in FormBaseTag, the setXXX methods were rewritten to just make
internal note of the value XXX, while initialiseState() would draw the
conclusions, and make internal fields, call other objects, etc. 

In InputTag(), there was a method setType() which made a call to FormBaseTag,
which was invalid if done during analysis (when initialiseState is not called on
the object). 


/usr/local/cvsroot/makumba/classes/org/makumba/view/jsptaglib/InputTag.java,v 
<--  InputTag.java
new revision: 2.4; previous revision: 2.3

**Comment by ghost** [comment]
**Comment by fred on 15 Nov 2010 16:58 UTC**
(In [602]) fix bug 654 and bug 583: bad tag life cycle for pooling
